product_multi_image/models/product_product.py

```diff
@@ -0,0 +1,116 @@
+        for product in self:
+            product._inverse_main_image(product.image_main)
+
+    @api.multi
+    def _inverse_main_image_medium(self):
+        for product in self:
+            product._inverse_main_image(product.image_main_medium)
+
+    @api.multi
+    def _inverse_main_image_small(self):
+        for product in self:
+            product._inverse_main_image(product.image_main_small)
+
+    @api.multi
+    @api.depends('product_tmpl_id', 'product_tmpl_id.image_ids',
+                 'product_tmpl_id.image_ids.product_variant_ids')
+    def _compute_image_ids(self):
+        for product in self:
+            images = product.product_tmpl_id.image_ids.filtered(
+                lambda x: (not x.product_variant_ids or
+                           product.id in x.product_variant_ids.ids))
+            product.image_ids = [(6, 0, images.ids)]
+
+    @api.multi
+    def _inverse_image_ids(self):
+        for product in self:
+            # Remember the list of images that were before changes
+            previous_images = product.product_tmpl_id.image_ids.filtered(
+                lambda x: (not x.product_variant_ids or
+                           product.id in x.product_variant_ids.ids))
+            for image in product.image_ids:
+                if isinstance(image.id, models.NewId):
+                    # Image added
+                    image.owner_id = product.product_tmpl_id.id
+                    image.owner_model = "product.template"
+                    image.product_variant_ids = [(6, 0, product.ids)]
+                    image.create(image._convert_to_write(image._cache))
+                else:
+                    previous_images -= image
+                    # Update existing records
+                    image.write(image._convert_to_write(image._cache))
+            for image in previous_images:
+                # Images removed
+                if not image.product_variant_ids:
+                    variants = product.product_tmpl_id.product_variant_ids
+                else:
+                    variants = image.product_variant_ids
+                variants -= product
+                if not variants:
+                    # Remove the image, as there's no variant that contains it
+                    image.unlink()
+                else:
+                    # Leave the images for the rest of the variants
+                    image.product_variant_ids = variants.ids
+
+    @api.multi
+    @api.depends('image_ids', 'product_tmpl_id.image_ids',
+                 'product_tmpl_id.image_ids.product_variant_ids')
+    def _get_multi_image(self):
+        """Needed for changing dependencies in this class."""
+        super(ProductProduct, self)._get_multi_image()
+
+
+class ProductProductOld(orm.Model):
+    """It is needed to use v7 api here because core model fields use the
+    ``multi`` attribute, that has no equivalent in v8, and it needs to be
+    disabled or bad things will happen.
```

Comment:
I think that a link to https://github.com/odoo/odoo/issues/10799 would fit here.